lib/compiler_transforms.ts

```diff
@@ -7,23 +7,48 @@ export const transformImportMeta: ts.TransformerFactory<ts.SourceFile> = (
   context,
 ) => {
   const factory = context.factory;
+  const compilerModule = context.getCompilerOptions().module;
+  const isScriptModule = compilerModule === ts.ModuleKind.CommonJS ||
+    compilerModule === ts.ModuleKind.UMD;
 
   return (sourceFile) => ts.visitEachChild(sourceFile, visitNode, context);
 
   function visitNode(node: ts.Node): ts.Node {
-    // find `import.meta.url`
+    let resolveArgs: undefined | ts.NodeArray<ts.Expression>;
+    let originNode: undefined | ts.Node;
+    if (ts.isCallExpression(node) && node.arguments.length === 1) {
+      resolveArgs = node.arguments;
+      originNode = node;
+      node = node.expression;
+    }
+    // find `import.meta.url` or `import.meta.main` or `import.meta.resolve`
     if (
       ts.isPropertyAccessExpression(node) &&
       ts.isMetaProperty(node.expression) &&
       node.expression.keywordToken === ts.SyntaxKind.ImportKeyword &&
       ts.isIdentifier(node.name)
     ) {
-      if (node.name.escapedText === "url") {
+      if (node.name.escapedText === "url" && isScriptModule) {
         return getReplacementImportMetaUrl();
       } else if (node.name.escapedText === "main") {
-        return getReplacementImportMetaMain();
+        if (isScriptModule) {
+          return getReplacementImportMetaMain();
+        } else {
+          return getReplacementImportMetaMainEsm();
+        }
+      } else if (
+        node.name.escapedText === "resolve" && resolveArgs !== undefined
+      ) {
+        return ts.visitEachChild(
+          getReplacementImportMetaResolve(resolveArgs),
+          visitNode,
+          context,
+        );
       }
     }
+    if (originNode !== undefined) {
+      node = originNode;
```

Comment:
This PR looks very good other than this state being maintained like this, which is a bit confusing. I'm going to push a change.